File: lib/go/TelepactSchemaParseError.go
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

package telepact

import (
	"fmt"

	"github.com/telepact/telepact/lib/go/internal/schema"
)

// TelepactSchemaParseError indicates a failure to parse a Telepact schema definition.
type TelepactSchemaParseError struct {
	SchemaParseFailures           []*schema.SchemaParseFailure
	SchemaParseFailuresPseudoJSON any
	message                       string
}

// NewTelepactSchemaParseError constructs a TelepactSchemaParseError with the supplied failures and schema documents.
func NewTelepactSchemaParseError(
	schemaParseFailures []*schema.SchemaParseFailure,
	telepactSchemaDocumentNamesToJSON map[string]string,
) *TelepactSchemaParseError {
	pseudoJSON := schema.MapSchemaParseFailuresToPseudoJSON(schemaParseFailures, telepactSchemaDocumentNamesToJSON)

	return &TelepactSchemaParseError{
		SchemaParseFailures:           schemaParseFailures,
		SchemaParseFailuresPseudoJSON: pseudoJSON,
		message:                       fmt.Sprintf("%v", pseudoJSON),
	}
}

// Error implements the error interface.
func (e *TelepactSchemaParseError) Error() string {
	if e == nil {
		return ""
	}
	return e.message
}

File: lib/ts/src/TelepactSchemaParseError.ts
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

import { SchemaParseFailure } from './internal/schema/SchemaParseFailure.js';
import { mapSchemaParseFailuresToPseudoJson } from './internal/schema/MapSchemaParseFailuresToPseudoJson.js';

export class TelepactSchemaParseError extends Error {
    /**
     * Indicates failure to parse a telepact Schema.
     */

    public schemaParseFailures: SchemaParseFailure[];
    public schemaParseFailuresPseudoJson: any;

    constructor(schemaParseFailures: SchemaParseFailure[], documentNamesToJson: Record<string, string>, cause?: Error) {
        super(JSON.stringify(mapSchemaParseFailuresToPseudoJson(schemaParseFailures, documentNamesToJson), null, 2), {
            cause: cause,
        });
        this.schemaParseFailures = schemaParseFailures;
        this.schemaParseFailuresPseudoJson = mapSchemaParseFailuresToPseudoJson(
            schemaParseFailures,
            documentNamesToJson,
        );
    }
}
